### PR TITLE
Issue warning on recursive shared example declaration

### LINKED ISSUE #2355: Shared Examples should raise error when included recursively
I just spent a few minutes wondering why this doesn't work

    shared_examples_for 'test' do
      # tons of code

      include_examples_for 'test'
    end

This does nothing. It should raise an error with message something like `can't include shared examples recursively (hint: move include_examples 'test' out of block shared_examples_for 'test')`

### PR DESCRIPTION
**Pull request by JonRowe**
Fixes #2355

### CODE REVIEW

File: lib/rspec/core/example_group.rb [hunk 1]
```diff
@@ -769,6 +769,11 @@ def initialize(shared_group_name, inclusion_location)
         @inclusion_location = inclusion_location
       end
 
+      # @private
+      def ==(other)
+        @shared_group_name == other.shared_group_name
+      end
```
Comment (JonRowe) [inline lib/rspec/core/example_group.rb]: This is just a short cut so we can check if theres an existing group with that name in our current call stack, whilst we do allow groups to share the same name, they will never be in the same call stack.

Comment (JonRowe) [inline lib/rspec/core/example_group.rb]: This is just a short cut so we can check if theres an existing group with that name in our current call stack, whilst we do allow groups to share the same name, they will never be in the same call stack.

Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: > whilst we do allow groups to share the same name, they will never be in the same call stack.

I get that, but we don't know that only the recursion check will call `==` in the future.  If we're going to define a `==` method, I want it to have the correct equality semantics of this class.

Instead, WDYT about removing this method and doing this below?

``` ruby
def self.with_frame(name, location)
  current_stack = shared_example_group_inclusions
  if current_stack.any? { |shared_group| shared_group.name == name }
    raise ArgumentError, "can't include shared examples recursively"
  else
    current_stack << new(name, location)
    yield
  end
ensure
  current_stack.pop
end
```

Seems just as simple.  Does not rely on knowledge that `==` will be called (as @mrageh asked below).  And avoids defining a `==` method with semantics that don't actually match equality semantics.

**Pull request merged**